File: 프로그래머스/무인도 여행.py
from collections import deque


def solution(maps):
    answer = []
    col = len(maps[0])
    row = len(maps)
    visited = [[False] * col for _ in range(row)]
    dq = deque([])

    def bfs(y, x, row, col):
        dy = [0, 0, -1, 1]
        dx = [-1, 1, 0, 0]
        sum = 0
        queue = deque([(y, x)])
        visited[y][x] = True

        while queue:
            cur_y, cur_x = queue.popleft()
            sum += int(maps[cur_y][cur_x])
            for i in range(4):
                next_y = cur_y + dy[i]
                next_x = cur_x + dx[i]

                if 0 <= next_y < row and 0 <= next_x < col and maps[next_y][next_x] != "X" and not visited[next_y][
                    next_x]:
                    queue.append((next_y, next_x))
                    visited[next_y][next_x] = True

        answer.append(sum)

    for y in range(row):
        for x in range(col):
            if maps[y][x] != "X" and not visited[y][x]:
                bfs(y, x, row, col)

    return sorted(answer) if len(answer) else [-1]

File: 프로그래머스/무인도 여행.js
function solution(maps) {
  var answer = [];
  const N = maps.length;
  const M = maps[0].length;
  const isVisited = Array.from({length: N}, () => new Array(M).fill(false));

  for (let row = 0; row < N; row++) {
    for (let col = 0; col < M; col++) {
      if (!isVisited[row][col] && maps[row][col] !== "X") {
        let result = dfs(row, col, maps, isVisited);
        if (result !== 0) answer.push(result);
      }
    }
  }

  answer.sort((a, b) => a - b);
  return answer.length === 0 ? [-1] : answer;
}

function dfs(y, x, maps, isVisited) {
  const vectors = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];
  const stack = [[y, x]];
  isVisited[y][x] = true;
  const N = maps.length;
  const M = maps[0].length;
  let result = Number(maps[y][x]);

  while (stack.length > 0) {
    let [curY, curX] = stack.pop();

    for (let [dy, dx] of vectors) {
      let nextY = curY + dy;
      let nextX = curX + dx;

      if (!(0 <= nextY && nextY < N && 0 <= nextX && nextX < M)) continue;
      if (maps[nextY][nextX] === "X") continue;
      if (isVisited[nextY][nextX]) continue;

      const value = Number(maps[nextY][nextX]);
      stack.push([nextY, nextX]);
      result += value;
      isVisited[nextY][nextX] = true;
    }
  }

  return result;
}
